### ISSUE TITLE
Files disappear sometimes in volume mounts

### ISSUE THREAD

**Issue by pbiggar** [moby/moby#41750]
---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------
**Description**

I'm not sure how to describe this issue that I'm having, but roughly speaking, I frequently find that directories in volumes either:
- a) vanish
- b) "escape" the container

I do not have a repro, and I honestly have no idea where to even start with this. It happens rarely enough that I've been unable to learn anything about it, and have instead papered over the problem.

It happens to me probably a few times a week and has for several years. It also happened to other employees.

For a) the specific thing that happens is that the directory, within the container, is empty. I have build output that shows it was just built, but there is no data there.

For b), sometimes I would start the container and discovered a ton of "build context". Upon investigating, it would be a directory like _build, which inexplicably had lots of build files in it on the host. The build files should not leave the container (and the build tools are not installed on the host so they must have gotten there by "leaking" from the container).

I describe these problems together as it feels they are related.

I'm using Docker Desktop, currently 2.5.01 (49550), though this has happened many times over many different versions.

This is my [Dockerfile](https://github.com/darklang/dark/blob/main/dockerfile-repo/Dockerfile), and I run Docker by calling [this script](https://github.com/darklang/dark/blob/main/scripts/builder).

Some relevant lines are [here](https://github.com/darklang/dark/blob/main/scripts/builder#L17), where I delete directories that have "leaked", and [here](https://github.com/darklang/dark/blob/main/scripts/builder#L73), where I set the mounts.

I mount `/home/dark/app` as a bind mount, then multiple directories within that are mounted as volumes, to avoid the performance penalty of syncing the build files to the host.

These are all build directories, and they have lots of traffic.

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Cloud integration: 1.0.2
 Version:           19.03.13
 API version:       1.40
 Go version:        go1.13.15
 Git commit:        4484c46d9d
 Built:             Wed Sep 16 16:58:31 2020
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.13
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       4484c46d9d
  Built:            Wed Sep 16 17:07:04 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.3.7
  GitCommit:        8fba4e9a7d01810a393d5d25a3621dc101981175
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683```

**Output of `docker info`:**

```
Client:
 Debug Mode: false
 Plugins:
  scan: Docker Scan (Docker Inc., v0.3.4)

Server:
 Containers: 2
  Running: 2
  Paused: 0
  Stopped: 0
 Images: 80
 Server Version: 19.03.13
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 8fba4e9a7d01810a393d5d25a3621dc101981175
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.4.39-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 23.49GiB
 Name: docker-desktop
 ID: WIVR:IVBV:YUTO:FPL2:AKMC:VCAH:2NXG:6CIS:P4T5:UCFM:W7LD:B4V7
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: gateway.docker.internal:3128
 HTTPS Proxy: gateway.docker.internal:3129
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
MacOS Catalina, Version 10.15.6. Intel core i9

**Comment by pbiggar** [comment]
Just wanted to follow up to say my volumes getting wiped is happening more and more. Dozens of times a day. It's completely destroying my productivity, between having to repeatedly rerun builds and my editor intellisense being broken.

If anyone has a workaround, that would be extremely appreciated!

**Comment by thaJeztah** [comment]
I think @djs55 is on PTO for the Holidays, but let me check with him on our Slack.


From https://github.com/darklang/dark/blob/main/scripts/builder#L17

> Sometimes these dirs leak out of the container, causing confusion and slowness

IIUC, you mean that files that _should be_ written to a volume that's mounted in the container, are instead written to the bind-mounted `/home/dark/app`, correct?

My main suspect would be either a race-condition (mounts applied in the incorrect order, therefore a bind-mount masking a volume-mount (resulting in those files to be written to the host, and files "disappearing" because the container now looks for those paths on the host, which may be empty)

This _could_ be related to how Docker Desktop handles file-paths; IIRC, it looks for paths that were mounted from the host, and sets up syncing for those paths (but @djs55 has more insight into those internals).

**Comment by djs55** [comment]
Have you tried a more recent Desktop version? There have been improvements to the filesharing implementation recently (in "gRPC FUSE" mode, not with `osxfs`) which might affect this. Even if the problem is still there, it would be useful to know as this would help focus the investigation.

**Comment by pbiggar** [comment]
> I think @djs55 is on PTO for the Holidays, but let me check with him on our Slack.
> 
> From https://github.com/darklang/dark/blob/main/scripts/builder#L17
> 
> > Sometimes these dirs leak out of the container, causing confusion and slowness
> 
> IIUC, you mean that files that _should be_ written to a volume that's mounted in the container, are instead written to the bind-mounted `/home/dark/app`, correct?

Yes, that's correct. However at the moment, it's more common that they just vanish. Like suddenly I get a compiler error, and when I `ls` the directory in the container it's empty.

Note: I use a long running container for my main "build server", which watches the host and runs build commands on file changes. I also simultaneously run a VS code "remote container", using the same image and some of the same volumes (some of the volumes which vanish are mounted in both running containers, some are only mounted on the build server).

**Comment by pbiggar** [comment]
> Have you tried a more recent Desktop version? There have been improvements to the filesharing implementation recently (in "gRPC FUSE" mode, not with `osxfs`) which might affect this. Even if the problem is still there, it would be useful to know as this would help focus the investigation.

OK, I'll try this. I recall I tried it once and it didn't improve, but I will try it again. Thanks!
